core/src/main/java/com/scalar/db/transaction/consensuscommit/CommitHandler.java

```diff
@@ -106,42 +106,58 @@ private void waitBeforePreparationSnapshotHookFuture(
     }
   }
 
-  public void commit(Snapshot snapshot) throws CommitException, UnknownTransactionStatusException {
+  public void commit(Snapshot snapshot, boolean readOnly)
+      throws CommitException, UnknownTransactionStatusException {
+    boolean hasNoWritesAndDeletesInSnapshot = readOnly || snapshot.hasNoWritesAndDeletes();
+
     Optional<Future<Void>> snapshotHookFuture = invokeBeforePreparationSnapshotHook(snapshot);
-    try {
-      prepare(snapshot);
-    } catch (PreparationException e) {
-      safelyCallOnFailureBeforeCommit(snapshot);
-      abortState(snapshot.getId());
-      rollbackRecords(snapshot);
-      if (e instanceof PreparationConflictException) {
-        throw new CommitConflictException(e.getMessage(), e, e.getTransactionId().orElse(null));
+
+    if (!hasNoWritesAndDeletesInSnapshot) {
```

Comment:
I think double negative is a bit hard to read.
Since all the usage of `hasNoWritesAndDeletesInSnapshot` is with negation (!), how about making the method like `hasSomeWritesAndDeletesInSnapshot` so that we don't have to negate all the if clauses?

Reply:
Fixed in 6c6299ddb11bc046d23766ef6b544dcbd7a3dc0c. Thanks!